src/bin/tree-sitter-graph/util.rs

```diff
@@ -0,0 +1,84 @@
+// -*- coding: utf-8 -*-
+// ------------------------------------------------------------------------------------------------
+// Copyright © 2022, tree-sitter authors.
+// Licensed under either of Apache License, Version 2.0, or MIT license, at your option.
+// Please see the LICENSE-APACHE or LICENSE-MIT files in this distribution for license details.
+// ------------------------------------------------------------------------------------------------
+
+use tree_sitter::Tree;
+
+#[derive(Debug, Default)]
+pub(crate) struct ParseErrors {
+    errors: Vec<(usize, usize, Option<String>)>,
+}
+
+impl ParseErrors {
+    pub(crate) fn from_tree(tree: &Tree, source: &str) -> Self {
+        let mut errors = ParseErrors { errors: Vec::new() };
+        // do not walk the tree unless there actually are errors
+        if !tree.root_node().has_error() {
+            return errors;
+        }
+
+        let mut cursor = tree.walk();
+        let mut did_visit_children = false;
+        loop {
+            let node = cursor.node();
+            if node.is_error() || node.is_missing() {
+                errors.add(node, source);
+                did_visit_children = true;
+            }
+            if did_visit_children {
+                if cursor.goto_next_sibling() {
+                    did_visit_children = false;
+                } else if cursor.goto_parent() {
+                    did_visit_children = true;
+                } else {
+                    break;
+                }
+            } else {
+                if cursor.goto_first_child() {
+                    did_visit_children = false;
+                } else {
+                    did_visit_children = true;
+                }
+            }
+        }
+        cursor.reset(tree.root_node());
+
+        errors
+    }
+
+    fn add(&mut self, node: tree_sitter::Node, source: &str) {
+        let start = node.start_position();
+        let node_source = if !node.byte_range().is_empty() {
+            Some(source[node.byte_range()].to_string())
+        } else {
+            None
+        };
+        self.errors.push((start.row, start.column, node_source));
+    }
+
+    pub(crate) fn is_empty(&self) -> bool {
+        self.errors.is_empty()
+    }
+}
```

Comment:
It would be nice to have a way to access the errors, beyond just displaying them or checking that there are none. 

The way I currently handle errors is to recognise them in a stanza, which is a bit awkward (since you can't actually query for `ERROR` nodes). However, it occurs to me that if I could access the list of errors, then I could just produce the relevant graph (consisting of a single "syntax error" node) directly, and do away with needlessly running the `tree-sitter-graph` machinery. (And in fact, I would be happy with just the _first_ syntax error in the file.)